### PR TITLE
Add handling for next parameter

### LINKED ISSUE #2385: Handle links inside dashboard when user is logged out
This information is in the `next` query parameter but we don't use when the user logs in via edX. Maybe there's something OAuth-related we can use to preserve this part after login?

### PR DESCRIPTION
**Pull request by noisecapella**
#### What are the relevant tickets?
Fixes #2385 

#### What's this PR do?
In the homepage the `next` query parameter is set in the link to login to edx. If the `next` query parameter is set for the link it is used as the redirect URI after the signin is completed. Since Django sets `?next` if the user attempts to go somewhere that requires a login, this should allow the user to just click the login link to get back where they were, including all relevant query parameters.

#### How should this be manually tested?
Log out, then go to `/dashboard/?xy=z`. You should be redirected to the home page. Click the login link to log in, then you should be redirected to edX and be directed back to end up at `/dashboard/?xy=z`.

### CODE REVIEW

File: backends/pipeline_api_test.py [hunk 1]
```diff
@@ -244,6 +245,19 @@ def test_login_redirect_learners(self):
...
+        """
+        A student should be directed to what's in the next query parameter
+        """
+        student = UserFactory.create()
+        session_set = mock.Mock()
+        next_url = "/x/y?a=bc"
+        load = mock.Mock(return_value={"next": next_url})
+        backend = edxorg.EdxOrgOAuth2(strategy=mock.Mock(session_set=session_set, session=mock.Mock(load=load)))
+        pipeline_api.update_profile_from_edx(backend, student, {}, False)
+        session_set.assert_called_with('next', next_url)
```
Comment (singingwolfboy) [inline backends/pipeline_api_test.py]: It's hard to follow the chain of mocks here. I'd suggest doing this, instead:

```python
student = UserFactory.create()
next_url = "/x/y?a=bc"
mock_strategy = mock.Mock()
mock_strategy.session.load.return_value = {"next": next_url}
backend = edxorg.EdxOrgOAuth2(strategy=mock_strategy)
pipeline_api.update_profile_from_edx(backend, student, {}, False)
mock_strategy.session_set.assert_called_with('next', next_url)
```

File: static/js/lib/api_test.js [hunk 2]
```diff
@@ -580,7 +573,7 @@ describe('api', function() {
           });
 
           return assert.isRejected(fetchJSONWithCSRF('/url', {}, true)).then(() => {
-            assert.equal(savedWindowLocation, '/login/edxorg/');
+            assert.isTrue(window.location.toString().includes('/login/edxorg/'));
```
Comment (singingwolfboy) [inline static/js/lib/api_test.js]: [`assert.include`](http://chaijs.com/api/assert/#method_include) would be clearer, and gets you a nicer error message:
```js
assert.include(window.location.toString(), '/login/edxorg/');
```

File: static/js/containers/SignupDialog_test.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    );
+  };
+
+  it('has a login link which uses the next query param', () => {
+    let queryParams = "?next=b";
+    window.location = `http://fake/${queryParams}`;
+    helper.store.dispatch(setDialogVisibility(true));
+    renderDialog();
+
+    let link = document.body.querySelector(".signup-dialog").querySelector("a");
```
Comment (singingwolfboy) [inline static/js/containers/SignupDialog_test.js]: Why not combine the two query selectors?
```js
let link = document.body.querySelector(".signup-dialog a");
```

**Comment by noisecapella** [comment]
Comments addressed

File: backends/pipeline_api_test.py [hunk 4]
```diff
@@ -232,7 +232,8 @@ def test_login_redirect_dashboard(self):
         """
         student = UserFactory.create()
         session_set = mock.Mock()
-        backend = edxorg.EdxOrgOAuth2(strategy=mock.Mock(session_set=session_set))
+        load = mock.Mock(return_value={})
+        backend = edxorg.EdxOrgOAuth2(strategy=mock.Mock(session_set=session_set, session=mock.Mock(load=load)))
```
Comment (singingwolfboy) [inline backends/pipeline_api_test.py]: Can you improve the mocks for this test, as well?

```python
student = UserFactory.create()
mock_strategy = mock.Mock()
mock_strategy.session.load.return_value = {}
backend = edxorg.EdxOrgOAuth2(strategy=mock_strategy)
pipeline_api.update_profile_from_edx(backend, student, {}, False)
mock_strategy.session_set.assert_called_with('next', '/dashboard')
```

File: backends/pipeline_api_test.py [hunk 5]
```diff
@@ -244,6 +245,19 @@ def test_login_redirect_learners(self):
             user = UserFactory.create()
             Role.objects.create(user=user, role=role, program=ProgramFactory.create())
             session_set = mock.Mock()
-            backend = edxorg.EdxOrgOAuth2(strategy=mock.Mock(session_set=session_set))
+            load = mock.Mock(return_value={})
+            backend = edxorg.EdxOrgOAuth2(strategy=mock.Mock(session_set=session_set, session=mock.Mock(load=load)))
```
Comment (singingwolfboy) [inline backends/pipeline_api_test.py]: Can you improve the mocks for this test, as well?
```python
user = UserFactory.create()
Role.objects.create(user=user, role=role, program=ProgramFactory.create())
mock_strategy = mock.Mock()
mock_strategy.session.load.return_value = {}
backend = edxorg.EdxOrgOAuth2(strategy=mock_strategy)
pipeline_api.update_profile_from_edx(backend, student, {}, False)
mock_strategy.session_set.assert_called_with('next', '/learners')
```
